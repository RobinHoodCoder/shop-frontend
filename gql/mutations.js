import gql from 'graphql-tag';

export const M_CREATE_PRODUCT = gql`
    mutation CREATE_PRODUCT_MUTATION (
        $name: String!
        $description: String!
        $price: Int!
        $image: Upload
    ) {
        createProduct(
            data: {
                name: $name
                description: $description
                price: $price
                status: "AVAILABLE"
                photo: {
                    create: {
                        image: $image,
                        altText: $name
                    }
                }
            }
        ) {
            id
            price
            description
            name
            id
        }
    }
`;

export const M_UPDATE_PRODUCT = gql`
  mutation UPDATE_PRODUCT (
    $id: ID!
    $name: String!
    $description: String!
    $price: Int!
#    $image: Upload
  ) {
    updateProduct(
      id: $id
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
#        photo: {
#          create: {
#            image: $image,
#            altText: $name
#          }
#        }
      }
    ) {
      id
      price
      description
      name
      id
    }
  }
`;

export const M_DELETE_PRODUCT = gql`
  mutation DELETE_PRODUCT (
    $id: ID!
  ) {
    deleteProduct(
      id: $id
    ) {
      name
      id
    }
  }
`;

export const M_LOGIN = gql`
    mutation LOGIN (
        $email: String!
        $password: String!
    ) {
        authenticateUserWithPassword(
            email: $email
            password: $password
        ) {
           ...on UserAuthenticationWithPasswordSuccess {
             item {
                 email
                 name
             }
           }
           ...on UserAuthenticationWithPasswordFailure {
            code
            message
           }
        }
    }
`;

export const M_LOGOUT = gql`
  mutation LOGOUT {
    endSession
  }
`;


export const M_SIGNUP = gql`
  mutation SIGNUP_MUTATION (
    $email: String!
    $password: String!
    $name: String!
  ) {
    createUser(
    data: {
      name: $name
      email: $email
      password: $password
    }
    ) {
        email
        id
        name
    }
  }
`;

export const M_REQUEST_RESET_PASSWORD = gql`
  mutation REQUEST_RESET_PASSWORD_MUTATION (
    $email: String!
  ) {
    sendUserPasswordResetLink(
      email: $email
    ) {
      code
      message
    }
  }
`;

export const M_RESET_PASSWORD = gql`
    mutation RESET_PASSWORD_MUTATION (
        $email: String!
        $token: String!
        $password: String!
    ) {
        redeemUserPasswordResetToken(
            email: $email
            token: $token
            password: $password
        ) {
            code
            message
        }
    }
`;

export const M_ADD_TO_CART = gql`
    mutation ADD_TO_CART_MUTATION($id: ID!) {
        addToCart(productId: $id) {
            id
        }
    }
`;

export const M_REMOVE_FROM_CART = gql`
    mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
        deleteCartItem(id: $id) {
            id
        }
    }
`;

export const M_CREATE_ORDER = gql`
    mutation M_CREATE_ORDER($token: String!) {
        checkout(token: $token) {
            id
            charge
            total
            items {
                id
                name
            }
        }
    }
`;
