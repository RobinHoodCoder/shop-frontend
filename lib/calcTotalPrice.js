import formatMoney from './formatMoney';

const calcTotalPrice = (cart) => {
  return cart.reduce((currentOutput, cartItem) => {
    console.log(currentOutput, cartItem);
    const { product } = cartItem || {};
    if (!product) { // check if it's available because product can be deleted
      return currentOutput;
    }

    if (!product.price) {
      return currentOutput;
    }
    return currentOutput + (cartItem.quantity * product.price);
  }, 0);
};

export const formattedTotalPrice = (cart) => {
  return formatMoney(
    calcTotalPrice(cart)
  );
};

export default calcTotalPrice;
