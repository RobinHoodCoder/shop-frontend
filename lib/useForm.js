import { useEffect, useState } from 'react';

export const useForm = (initial = {}) => {
  const [formValues, setInputs] = useState(initial);
  const handleChange = (e) => {
    const { value, name, type, files } = e.target;

    let outVal = '';

    switch (type) {
    case 'number' :
      outVal = !!value ? parseInt(value, 10) : 0;
      break;
    case 'file' :
      // console.log(value);
      [outVal] = files;
      break;
    default:
      outVal = value;
    }

    setInputs((prevState) => {
      return {
        ...prevState,
        [name]: outVal,
      };
    });
  };
  const resetForm = (e) => {
    e.preventDefault();
    setInputs(initial);
  };
  const clearForm = (e) => {
    e.preventDefault();
    const blankArr = Object.entries(initial).map(([key, _]) => {
      return [key, ''];
    });
    const blankObj = Object.fromEntries(blankArr);
    setInputs(blankObj);
  };

  useEffect(() => {

  }, []);

  return { formValues, handleChange, resetForm, clearForm };
};
