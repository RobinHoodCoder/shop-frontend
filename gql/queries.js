import gql from 'graphql-tag';

export const Q_ALL_PRODUCTS = gql`
    query ALL_PRODUCTS ( $first: Int = 1, $skip: Int = 0, ) 
    {
    allProducts (
     first: $first
     skip: $skip
    ) {
        id
        name
        description
        price
        photo {
            altText
            id
            image
            {
                publicUrlTransformed
            }
        }
    }
}
`;
export const Q_CURRENT_USER = gql`
    query  {
    authenticatedItem {
        ...on User {
                id
                email
                name
                # Query the curruser once we have it
            }
        }
    }
`;

export const Q_SINGLE_PRODUCT = gql`
    query SINGLE_PRODUCT (
       $id: ID!
    ) 
    {
        Product(where: {id: $id}) {
            id
            name
            description
            price
            photo {
                altText
                id
                image
                {
                    publicUrlTransformed
                }
            }
        }
    }
`;
export const Q_PAGINATION = gql`
    query PAGINATION_QUERY
    {
        _allProductsMeta {
            count
        }
    }
`;
