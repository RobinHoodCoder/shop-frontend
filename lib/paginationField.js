import gql from 'graphql-tag';

const paginationField = () => {
  console.log('paginationField');
  return {
    keyArgs: false,
    read(
      existing = [], // e.g. }]
      {
        args, // gimme the first x things  and skip 2
        cache, // reference to the cache from Apollo
      } // e.g. :
    ) {
      const { skip, first } = args;
      // Read number of items from cache

      const data = cache.readQuery({
        Q_PAGINATION: {
          skip,
          first,
        },
      });
      const count = data?._allProductsMeta?.count;

      const page = (skip / first) + 1;
      const pages = Math.ceil(count / first);
      // ask read func for the items

      // look for non-existing items
      const items = existing.slice(skip, skip + 1);

      if (items.length !== first) {
        // We dont have all the items, so we need to fetch them
        // console.log(existing, args, cache, data);
        return false;
      }
      // If there are items, first return from cache, and we don' t need to go to the network.
      if (items.length) {
        console.log(`There are ${items.length} items in the cache. we gonna send em to Apollo.`);
        return items;
      }
    },
    merge(newThings) {
      console.log(newThings);
      // Runs when Apollo client turns back from the network.
    },
  };
};

export default paginationField;
