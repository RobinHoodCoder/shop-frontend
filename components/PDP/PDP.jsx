import React from 'react';
import { useQuery } from '@apollo/client';
import { Q_SINGLE_PRODUCT } from '../../gql/queries';
import DisplayError from '../ErrorMessage';
import  Head from 'next/head';
import styled from 'styled-components';
import Price from '../Price/Price';
import TitleBlock from '../TitleBlock/TitleBlock';
import Link from 'next/link';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCoffee, faEdit } from '@fortawesome/free-solid-svg-icons';
import { AddToCartButton } from '../AddToCartButton/AddToCartButton';

const ProductStyles = styled.main`
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 200px;
  align-items: center;
  gap: 2rem;

  .content {
    padding: 0 1.5rem;
  }

  img {
    display: block;
    position: relative;
    width: 100%;
    object-fit: contain;
    max-height: 50vh;
  }
`;

const PDP = ({ id }) => {
  const { data = {}, loading, error } = useQuery(Q_SINGLE_PRODUCT, { variables: { id } });
  const { Product = {} }  = data;

  const { name, description, price, photo } = Product;


  return (
    <>
      <Head>
        <title>Sick fits | {name}</title>
      </Head>
      {!error ? (
        !loading && (
          <ProductStyles>
            <div className="img-container">
              <img src={photo?.image?.publicUrlTransformed} alt={photo.altText} />
            </div>
            <div className={'content'}>
              <TitleBlock
                name={name}
                description={description}
                price={price}
                centered
              />
              <Price amount={price}/>
              <Link href={{
                pathname: '/edit',
                query: {
                  id,
                },
              }}>
                <FontAwesomeIcon
                  size="2x"
                  icon={faEdit}
                />
              </Link>
              <AddToCartButton id={id}/>
            </div>
          </ProductStyles>
        )
      ) : (
        <DisplayError error={error}/>
      )}
    </>
  );
};

export default PDP;
