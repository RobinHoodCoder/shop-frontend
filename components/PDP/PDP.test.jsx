import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import { fakeItem } from '../../lib/testUtils';
import { Q_SINGLE_PRODUCT } from '../../gql/queries';
import PDP from './PDP';
import { GraphQLError } from 'graphql';

const product = fakeItem();

const mocks = [
  {
    // When someone requests this query and variable combo
    request: {
      query: Q_SINGLE_PRODUCT,
      variables: {
        id: '123',
      },
    },
    // Return this data
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe('<PDP/>', () => {
  it('renders with proper data', async () => {
    // We need to make some fake data
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <PDP id="123" />
      </MockedProvider>
    );
    // Wait for the test ID to show up
    await screen.findByTestId('price');
    expect(container)
      .toMatchSnapshot();
  });

  it('Errors out when an item is no found', async () => {
    const errorMock = [
      {
        request: {
          query: Q_SINGLE_PRODUCT,
          variables: {
            id: '123',
          },
        },
        result: {
          errors: [new GraphQLError('Error!')],
        },
        // error: new Error('Error!'),
      },
    ];
    const { container, debug } = render(
      <MockedProvider mocks={errorMock}>
        <PDP id="123" />
      </MockedProvider>
    );
    await screen.findByTestId('graphql-error');
    expect(container)
      .toHaveTextContent('Shoot!');
    expect(container)
      .toHaveTextContent('Item not found!!!');
  });
});
