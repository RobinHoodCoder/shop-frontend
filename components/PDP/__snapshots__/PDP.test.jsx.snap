// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PDP/> renders with proper data 1`] = `
<div>
  <main
    class="PDP__ProductStyles-sc-o4iv9c-0 bUuHYa"
  >
    <div
      class="img-container"
    >
      <img
        alt="dogs are best"
        src="dog.jpg"
      />
    </div>
    <div
      class="content"
    >
      <div
        class="TitleBlock__TitleBlockStyles-sc-1oa93hk-0 EzZnr"
      >
        <h3
          class="Title-sc-16nq74k-0 jKqyec"
        >
          <span>
            dogs are best
          </span>
        </h3>
        <p
          class="Description-sc-1etfdho-0 jCgNpW"
        >
          <span>
            dogs
          </span>
        </p>
      </div>
      <div
        data-testid="price"
      >
        <span
          class="Price__PriceStyles-sc-1j6li5c-0 ezLsZj"
        >
          â‚¬50
        </span>
      </div>
      <div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-edit fa-w-18 fa-2x "
          data-icon="edit"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
            fill="currentColor"
          />
        </svg>
      </div>
      <button
        type="button"
      >
        Add
         To Cart ðŸ›’
      </button>
    </div>
  </main>
</div>
`;
